src/librustc_typeck/check/mod.rs

```diff
@@ -2357,8 +2357,18 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
 
                 let mut adjustments = autoderef.adjust_steps(needs);
                 if let ty::TyRef(region, mt) = method.sig.inputs()[0].sty {
+                    let mutbl = match mt.mutbl {
+                        hir::MutImmutable => AutoBorrowMutability::Immutable,
+                        hir::MutMutable => AutoBorrowMutability::Mutable {
+                            // FIXME: arguably indexing is "just
+                            // another kind of call"; perhaps it would
+                            // be more consistent to allow two-phase
+                            // borrows for .index() receivers here.
+                            allow_two_phase_borrows: false,
```

Comment:
(I think the test illustrates the kind of code that is being filtered out by not allowing two-phase borrows for index operators... I'll leave another comment with my thoughts on that test itself.)